contracts/core/modules/IssuanceOrderModule.sol

```diff
@@ -405,48 +327,194 @@ contract IssuanceOrderModule is
      *
-     * @param  _order               IssuanceOrder object containing order params
-     * @param  _fillQuantity        Quantity of Set to be filled
-     * @param  _orderData           Bytestring encoding all exchange order data
+     * @param _orderData               Bytes array containing the exchange orders to execute
+     * @param _makerAddress            Issuance order maker address
+     * @param _makerTokenAddress       Address of maker token to use to execute exchange orders
+     * @param _fillQuantity            Quantity of Set to be filled
+     * @param _attemptedFillQuantity   Quantity of Set taker attempted to fill
+     * @return makerTokenUsed          Amount of maker token used to execute orders
      */
-    function settleOrder(
-        OrderLibrary.IssuanceOrder _order,
+    function executeExchangeOrders(
+        bytes _orderData,
+        address _makerAddress,
+        address _makerTokenAddress,
         uint256 _fillQuantity,
-        bytes _orderData
+        uint256 _attemptedFillQuantity
     )
         private
+        returns (uint256)
     {
-        // Declare IVault interface as variable
-        IVault vaultInstance = IVault(vault);
+        uint256 scannedBytes;
+        uint256 makerTokenUsed;
+        while (scannedBytes < _orderData.length) {
 
-        // Check to make sure open order amount equals _fillQuantity
-        uint256 closedOrderAmount = orderFills[_order.orderHash].add(orderCancels[_order.orderHash]);
+            // Parse next exchange header based on scannedBytes
+            ExchangeHeaderLibrary.ExchangeHeader memory header = ExchangeHeaderLibrary.parseExchangeHeader(
+                _orderData,
+                scannedBytes
+            );
 
-        // Open order amount is greater than or equal to closed order amount
-        require(
-            _order.quantity.sub(closedOrderAmount) >= _fillQuantity,
-            "IssuanceOrderModule.settleOrder: Fill amount exceeds order available quantity"
+            // Get exchange address from state mapping based on header exchange info
+            address exchangeWrapper = ICore(core).exchanges(header.exchange);
+
+            // Verify exchange address is registered
+            require(
+                exchangeWrapper != address(0),
+                "IssuanceOrderModule.executeExchangeOrders: Invalid or disabled Exchange address"
+            );
+
+            // Read the order body based on order data length info in header plus the length of the header (128)
+            uint256 exchangeDataLength = header.orderDataBytesLength.add(128);
+            bytes memory bodyData = LibBytes.slice(
+                _orderData,
+                scannedBytes.add(128),
+                scannedBytes.add(exchangeDataLength)
+            );
+
+            // Calculate amount of makerToken actually needed
+            uint256 neededMakerTokenAmount = OrderLibrary.getPartialAmount(
```

Comment:
What if `_fillQuantity == `_attemptedFillQuantity`? Will we still need to calculate partialAmount each time as well as in the exchange wrappers?

Reply:
Unfortunately yeah. It's not ideal, unless we want to add a lot of `if...else...` statements, which we can do.

Reply:
I don't have an answer.. but there might be an architectural / structural change that we can make to reduce the duplication of logic in calculating the partialAmount.

Reply:
Let's get this merged first, and then we can see what the invocation cost differences are with bypassing the calculations. Add a trello for this